chia/wallet/coin_selection.py

```diff
@@ -78,28 +78,31 @@ async def select_coins(
         log.debug(f"Selected all smaller coins because they equate to an exact match of the target.: {smaller_coins}")
         return set(smaller_coins)
     elif smaller_coin_sum < amount:
-        smallest_coin = select_smallest_coin_over_target(len(smaller_coins), valid_spendable_coins)
+        smallest_coin: Optional[Coin] = select_smallest_coin_over_target(amount, valid_spendable_coins)
+        assert smallest_coin is not None  # Since we know we have enough, there must be a larger coin
         log.debug(f"Selected closest greater coin: {smallest_coin.name()}")
         return {smallest_coin}
     elif smaller_coin_sum > amount:
-        coin_set = knapsack_coin_algorithm(smaller_coins, amount, max_coin_amount)
+        coin_set: Optional[Set[Coin]] = knapsack_coin_algorithm(smaller_coins, amount, max_coin_amount, max_num_coins)
         log.debug(f"Selected coins from knapsack algorithm: {coin_set}")
         if coin_set is None:
-            raise ValueError("Knapsack algorithm failed to find a solution.")
-        if len(coin_set) > max_num_coins:
-            coin = select_smallest_coin_over_target(len(smaller_coins), valid_spendable_coins)
-            if coin is None or coin.amount < amount:
-                raise ValueError(
-                    f"Transaction of {amount} mojo would use more than "
-                    f"{max_num_coins} coins. Try sending a smaller amount"
-                )
-            coin_set = {coin}
+            coin_set = sum_largest_coins(amount, smaller_coins)
+            if coin_set is None or len(coin_set) > max_num_coins:
+                greater_coin = select_smallest_coin_over_target(amount, valid_spendable_coins)
+                if greater_coin is None:
+                    raise ValueError(
+                        f"Transaction of {amount} mojo would use more than "
+                        f"{max_num_coins} coins. Try sending a smaller amount"
+                    )
+                coin_set = {greater_coin}
         return coin_set
     else:
         # if smaller_coin_sum == amount and len(smaller_coins) >= max_num_coins.
-        coin = select_smallest_coin_over_target(len(smaller_coins), valid_spendable_coins)
-        log.debug(f"Resorted to selecting smallest coin over target due to dust.: {coin}")
-        return {coin}
+        potential_large_coin: Optional[Coin] = select_smallest_coin_over_target(amount, valid_spendable_coins)
+        if potential_large_coin is None:
+            raise ValueError("Too many coins are required to make this transaction")

```

Comment:
In the interest of consistency - should this be the same error text as above

` f"Transaction of {amount} mojo would use more than {max_num_coins} coins. Try sending a smaller amount"`